From 12d29bf383b64c6ed874c57a60391970b60aef3f Mon Sep 17 00:00:00 2001
From: Ryan Walklin <ryan@testtoast.com>
Date: Wed, 5 Jun 2024 20:06:32 +1200
Subject: [PATCH] arm64: dts: allwinner: h616: Add TCON nodes to H616 DTSI

The Allwinner H616 has a display pipeline similar to other Allwinner
devices, specifically the A10. Not all output pins are exposed on all
package variants (for example only the H700 and T507 have LCD pins
exposed, but all variants support at least HDMI outputs connected to a
display engine via a TCON TOP and one or more timing controllers
(TCONs).

Add TCON nodes for the TOP, and the LCD and TV timing controllers. The
timing controllers are compatible with the existing R40 driver.

Signed-off-by: Ryan Walklin <ryan@testtoast.com>
---
 .../arm64/boot/dts/allwinner/sun50i-h616.dtsi | 132 ++++++++++++++++++
 1 file changed, 132 insertions(+)

diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h616.dtsi b/arch/arm64/boot/dts/allwinner/sun50i-h616.dtsi
index 18aae70d58ec..2c395b7d67a0 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h616.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h616.dtsi
@@ -784,6 +787,36 @@ ohci3: usb@5311400 {
 			status = "disabled";
 		};
 
+		tcon_lcd0: lcd-controller@6511000 {
+			compatible = "allwinner,sun8i-r40-tcon-lcd";
+			reg = <0x06511000 0x1000>;
+			interrupts = <GIC_SPI 64 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&ccu CLK_BUS_TCON_LCD0>, <&ccu CLK_TCON_LCD0>;
+			clock-names = "ahb", "tcon-ch0";
+			clock-output-names = "tcon-data-clock";
+			#clock-cells = <0>;
+			resets = <&ccu RST_BUS_TCON_LCD0>, <&ccu RST_BUS_TCON_LCD1>;
+			reset-names = "lcd", "lvds";
+			status = "disabled";
+
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				tcon_lcd0_in: port@0 {
+					reg = <0>;
+
+					tcon_lcd0_in_tcon_top_mixer0: endpoint {
+						remote-endpoint = <&tcon_top_mixer0_out_tcon_lcd0>;
+					};
+				};
+
+				tcon_lcd0_out: port@1 {
+					reg = <1>;
+				};
+			};
+		};
+
 		rtc: rtc@7000000 {
 			compatible = "allwinner,sun50i-h616-rtc";
 			reg = <0x07000000 0x400>;
-- 
2.43.4

--- a/arch/arm64/boot/dts/allwinner/sun50i-h616.dtsi	2024-06-10 07:31:45.664006072 +0200
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h616.dtsi	2024-06-10 07:34:27.419357090 +0200
@@ -1074,6 +1074,11 @@
 					#size-cells = <0>;
 					reg = <1>;
 
+					tcon_top_mixer0_out_tcon_lcd0: endpoint@0 {
+						reg = <0>;
+						remote-endpoint = <&tcon_lcd0_in_tcon_top_mixer0>;
+					};
+
 					tcon_top_mixer0_out_tcon_tv: endpoint@2 {
 						reg = <2>;
 						remote-endpoint = <&tcon_tv_in_tcon_top_mixer0>;
